### PR TITLE
Adds tests for snapshot atomicity and iteration

### PR DESCRIPTION
**Pull request by dylanlott**
### Description

Adds tests for snapshot atomicity and iteration.

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-postgres/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

**Comment by lovromazgon** [comment]
Tests are looking nice already! 👍

File: test/helper.go [hunk 1]
```diff
@@ -100,7 +100,9 @@ func SetupTestTable(ctx context.Context, t *testing.T, conn Querier) string {
 }
 
 func RandomIdentifier(t *testing.T) string {
-	return fmt.Sprintf("conduit_%v_%d", strings.ToLower(t.Name()), time.Now().UnixMicro()%1000)
+	return fmt.Sprintf("conduit_%v_%d",
+		strings.ReplaceAll(strings.ToLower(t.Name()), "/", "_"),
```
Comment (dylanlott) [inline test/helper.go:104]: Necessary to make test names with `/` in them work with Postgres table naming.

File: source/logrepl/snapshot_test.go [hunk 2]
```diff
@@ -24,28 +24,100 @@ import (
...
 	is.NoErr(err)
-	t.Cleanup(conn.Release)
+
+	// assert record does not appear in snapshot
+	for i := 0; i < 5; i++ {
+		r, err := s.Next(ctx)
+		if err != nil {
+			is.Equal(err, ErrSnapshotComplete)
+			is.Equal(r, sdk.Record{})
+			is.Equal(s.complete, true)
```
Comment (lovromazgon) [inline source/logrepl/snapshot_test.go:96]: Same comment about asserting private fields, all we care about is getting `ErrSnapshotComplete` from `Next`.

File: source/logrepl/snapshot_test.go [hunk 3]
```diff
@@ -24,28 +24,100 @@ import (
...
+		}
+	}
+}
+
+func TestFullIteration(t *testing.T) {
```
Comment (lovromazgon) [inline source/logrepl/snapshot_test.go:101]: Would be great to add a call to `Teardown` and assert it didn't fail after the iteration is done to test the complete lifecycle of the iterator.

File: source/logrepl/snapshot_test.go [hunk 4]
```diff
@@ -24,28 +24,100 @@ import (
...
-	ctx := context.Background()
+	type args struct {
+		ctx context.Context
+	}
+	tests := []struct {
+		name    string
+		setup   func(t *testing.T) *SnapshotIterator
+		args    args
+		wantErr bool
+		wanted  error
```
Comment (hariso) [inline source/logrepl/snapshot_test.go:45]: Would it be simpler to have only `wanted`?

File: source/logrepl/snapshot_test.go [hunk 5]
```diff
@@ -103,3 +161,25 @@ func createTestSnapshot(t *testing.T, pool *pgxpool.Pool) string {
...
 }
+
+// createTestSnapshotIterator creates a new test table, starts a snapshot
+// on it, then creates a test SnapshotIterator with the ID of that snapshot.
+// * It returns that SnapshotIterator and the string name of the test table.
+// * This function handles its own pooled connection cleanup, but _not_ the
+// SnapshotIterator's Teardown.
+func createTestSnapshotIterator(ctx context.Context, t *testing.T,
+	pool *pgxpool.Pool, snapshotName string, table string, columns []string,
+	key string) *SnapshotIterator {
```
Comment (lovromazgon) [inline source/logrepl/snapshot_test.go:172]: Nitpick: the last 4 params could be replaced by `SnapshotConfig`.

File: source/logrepl/snapshot_test.go [hunk 6]
```diff
@@ -103,3 +161,25 @@ func createTestSnapshot(t *testing.T, pool *pgxpool.Pool) string {
 
 	return name
 }
+
+// createTestSnapshotIterator creates a new test table, starts a snapshot
+// on it, then creates a test SnapshotIterator with the ID of that snapshot.
+// * It returns that SnapshotIterator and the string name of the test table.
+// * This function handles its own pooled connection cleanup, but _not_ the
+// SnapshotIterator's Teardown.
```
Comment (lovromazgon) [inline source/logrepl/snapshot_test.go:169]: The doc needs to be updated now.

**Review by lovromazgon — APPROVED** [review]
Approving already to not block this further, but I'd appreciate it if subtests were split into normal tests. Looking good otherwise 👍

File: source/logrepl/snapshot_test.go [hunk 7]
```diff
@@ -24,65 +24,123 @@ import (
...
 	})
-	is.NoErr(err)
-	t.Cleanup(conn.Release)
+}
 
-	now := time.Now()
-	rec, err := s.Next(ctx)
-	is.NoErr(err)
+func TestSnapshot(t *testing.T) {
+	is := is.New(t)
```
[the remark this answers is not in the document]
Comment (dylanlott) [reply source/logrepl/snapshot_test.go:66]: That's good to know 👍 Thanks!

File: source/logrepl/snapshot_test.go [hunk 8]
```diff
@@ -24,65 +24,123 @@ import (
...
+		is.True(errors.Is(err, ErrSnapshotInterrupt))
 	})
-	is.NoErr(err)
-	t.Cleanup(conn.Release)
+}
 
-	now := time.Now()
-	rec, err := s.Next(ctx)
-	is.NoErr(err)
+func TestSnapshot(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (dylanlott) [reply source/logrepl/snapshot_test.go]: I slightly disagree, they all test constraints of the snapshot's lifecycle, but I factored them back out to their own tests.

[the remark this answers is not in the document]
Comment (lovromazgon) [reply source/logrepl/snapshot_test.go]: To clarify what I meant - they have no setup or cleanup in common, they are pretty much independent of each other, that's why I think they should be separate tests (that also makes it easier to run each one separately). The fact that they are in this file already indicates that they test the snapshot iterator, no need to further group IMO. I would probably go with a subtest if there were some substantial setup/cleanup that we don't want to repeat for each test, or for table-driven tests.

**Pull request merged**